src/pyramid_view.cpp

```diff
@@ -0,0 +1,321 @@
+
+          TENSORSTORE_CHECK_OK_AND_ASSIGN(auto source, tensorstore::Open(
+            GetOmeTiffSpecToRead(image_coll_path + "/" + file_name),
+            tensorstore::OpenMode::open,
+            tensorstore::ReadWriteMode::read).result());
+          PLOG_INFO << "Opening " << file_name;
+          auto image_shape = source.domain().shape();
+          auto image_width = image_shape[4];
+          auto image_height = image_shape[3];
+          auto array = tensorstore::AllocateArray({
+              image_height,
+              image_width
+            }, tensorstore::c_order,
+            tensorstore::value_init, source.dtype());
+
+          // initiate a read
+          tensorstore::Read(source |
+            tensorstore::Dims(3).ClosedInterval(0, image_height - 1) |
+            tensorstore::Dims(4).ClosedInterval(0, image_width - 1),
+            array).value();
+
+          const auto & [x_grid, y_grid, c_grid] = location;
+
+          tensorstore::IndexTransform < > transform = tensorstore::IdentityTransform(dest.domain());
+          if (v == VisType::PCNG) {
+            transform = (std::move(transform) | tensorstore::Dims("z", "channel").IndexSlice({
+                0,
+                c_grid
+              }) |
+              tensorstore::Dims(y_dim).SizedInterval(y_grid * whole_image._chunk_size_y, image_height) |
+              tensorstore::Dims(x_dim).SizedInterval(x_grid * whole_image._chunk_size_x, image_width) |
+              tensorstore::Dims(x_dim, y_dim).Transpose({
+                y_dim,
+                x_dim
+              })).value();
+
+          } else if (v == VisType::NG_Zarr) {
+            transform = (std::move(transform) | tensorstore::Dims(c_dim).SizedInterval(c_grid, 1) |
+              tensorstore::Dims(y_dim).SizedInterval(y_grid * whole_image._chunk_size_y, image_height) |
+              tensorstore::Dims(x_dim).SizedInterval(x_grid * whole_image._chunk_size_x, image_width)).value();
+          } else if (v == VisType::Viv) {
+            transform = (std::move(transform) | tensorstore::Dims(c_dim).SizedInterval(c_grid, 1) |
+              tensorstore::Dims(y_dim).SizedInterval(y_grid * whole_image._chunk_size_y, image_height) |
+              tensorstore::Dims(x_dim).SizedInterval(x_grid * whole_image._chunk_size_x, image_width)).value();
+          }
+          tensorstore::Write(array, dest | transform).value();
+        });
+      }
+
+      th_pool.wait_for_tasks();
+    }
+    base_image = whole_image;
+  }
+
+  void PyramidView::ReAssembleBaseLevelWithNewMap(VisType v, const image_map& m, const std::string& output_path) {
+
+    auto [x_dim, y_dim, c_dim, num_dims] = GetZarrParams(v);
+
+    auto input_spec = [v, this]() {
+      if (v == VisType::NG_Zarr | v == VisType::Viv) {
+        return GetZarrSpecToRead(base_zarr_path);
+      } else { // this will probably never happen
+        return tensorstore::Spec();
```

Comment:
Could there be any dangers of returning a default Spec? The comment on the tensorstore code says "Constructs an invalid specification". Maybe just log a warning if this could cause issues later on in the code path?